pkg/proxy/client_conn.go

```diff
@@ -0,0 +1,260 @@
+	}
+}
+
+var _ ClientConn = (*clientConn)(nil)
+
+// newClientConn creates a new client connection.
+func newClientConn(
+	ctx context.Context,
+	logger *log.MOLogger,
+	conn goetty.IOSession,
+	mc clusterservice.MOCluster,
+	router Router,
+	tun *tunnel,
+) (ClientConn, error) {
+	host, _, err := net.SplitHostPort(conn.RemoteAddress())
+	if err != nil {
+		return nil, err
+	}
+	c := &clientConn{
+		ctx:       ctx,
+		log:       logger,
+		conn:      conn,
+		connID:    nextClientConnID(),
+		moCluster: mc,
+		router:    router,
+		tun:       tun,
+		originIP:  net.ParseIP(host),
+	}
+	fp := config.FrontendParameters{}
+	fp.SetDefaultValues()
+	c.mysqlProto = frontend.NewMysqlClientProtocol(c.connID, c.conn, 0, &fp)
+	return c, nil
+}
+
+// RawConn implements the ClientConn interface.
+func (c *clientConn) RawConn() net.Conn {
+	if c != nil {
+		return c.conn.RawConn()
+	}
+	return nil
+}
+
+// GetTenant implements the ClientConn interface.
+func (c *clientConn) GetTenant() Tenant {
+	if c != nil {
+		return c.account.tenant
+	}
+	return EmptyTenant
+}
+
+// SendErrToClient implements the ClientConn interface.
+func (c *clientConn) SendErrToClient(errMsg string) {
+	accErr := moerr.MysqlErrorMsgRefer[moerr.ER_ACCESS_DENIED_ERROR]
+	p := c.mysqlProto.MakeErrPayload(accErr.ErrorCode, accErr.SqlStates[0], errMsg)
+	if err := c.mysqlProto.WritePacket(p); err != nil {
+		c.log.Error("failed to send access error to client", zap.Error(err))
+	}
+}
+
+// BuildConnWithServer implements the ClientConn interface.
+func (c *clientConn) BuildConnWithServer(handshake bool) (ServerConn, error) {
+	if handshake {
+		// Step 1, proxy write initial handshake to client.
+		if err := c.writeInitialHandshake(); err != nil {
+			c.log.Error("failed to write Handshake packet", zap.Error(err))
+			return nil, err
+		}
+		// Step 2, client send handshake response, which is auth request,
+		// to proxy.
+		if err := c.handleHandshakeResp(); err != nil {
+			c.log.Error("failed to handle Handshake response", zap.Error(err))
+			return nil, err
+		}
+	}
+	// Step 3, proxy connects to a CN server to build connection.
+	conn, err := c.connectToBackend(handshake)
+	if err != nil {
+		c.log.Error("failed to connect to backend", zap.Error(err))
+		return nil, err
+	}
+	return conn, nil
+}
+
+// Close implements the ClientConn interface.
+func (c *clientConn) Close() error {
+	return nil
+}
+
+// connectToBackend connect to the real CN server.
+func (c *clientConn) connectToBackend(sendToClient bool) (ServerConn, error) {
+	// Testing path.
+	if c.testHelper.connectToBackend != nil {
+		return c.testHelper.connectToBackend()
+	}
+
+	if c.router == nil {
+		return nil, moerr.NewInternalErrorNoCtx("no router available")
+	}
+
+	// Select the best CN server from backend.
+	//
+	// NB: The selected CNServer must have label hash in it.
+	cn, err := c.router.Select(c.labelInfo)
+	if err != nil {
+		return nil, err
+	}
+	sc, r, err := c.router.Connect(cn, c.handshakePack, c.tun)
```

Comment:
I don't understand why this works, when `connectToBackend` is called with `sentToClient==false` (namely, transfer the tunnel), the clientConn use cached `HandshakePacket` to authenticate the CN server. However, MySQL protocol [use an encryption salt to prevent replay attack](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_authentication_methods_native_password_authentication.html). How can the auth response with the old salt can authenticate the new CN server?

Reply:
salt is hard code and need to be fixed, but have no plan for it.